Data access api README/Docs update (#744)
IMAP-Science-Operations-Center/imap_processing: docs/source/data-access-api/index.rst

Reviewer: Since this document is quite lengthy now and have several sections and subsections, I suggest to add a table of contents somewhere towards the top. (Could be done in a future PR if desired)

Reviewer: Since this document is quite lengthy now and have several sections and subsections, I suggest to add a table of contents somewhere towards the top. (Could be done in a future PR if desired)

Reviewer: Nevermind, Ana pointed out to me that there is a built-in table of contents on ReadTheDocs, which I totally missed!

===============

The `imap-data-access <https://github.com/IMAP-Science-Operations-Center/imap-data-access>`_

> repository provides programmatic access, and a command-line utility for
> interacting with the API. It is the preferred way to use the API.

Reviewer: ```suggestion
repository provides a command-line utility and python package for
interacting with the API programmatically. It is the preferred way to use the API.
```

> The SDC also provides a REST API that allows users to upload and download files as

Reviewer: I suggest to change this whole paragraph to something like this:

```suggestion
Users may also download, upload, and query via the REST API directly through the browser or `curl` commands. The [REST API Specification](link) section below describes the various endpoints that are supported and how to use them.
```

The API can be accessed from the following URL [WIP]: https://api.dev.imap-mission.com

Command Line Utility
====================
To Install

> ----------

Reviewer: ```suggestion
Command Line Utility
--------------------
To Install
^^^^^^^^^^
```

It looks like these headings (and throughout this document) are still messing up the main RtD page layout. I think these need to be changed to another level 'down' in markdown.

Author: Oh! Sorry, Okay. I think I understand what you are saying now! Should I do this with every heading?

Reviewer: Yes, except for the very first one (line 4), which is fine as is.

Run the following command to use the API CLI.

    .. code-block:: bash

        pip install imap-data-access

Base Command Arguments
----------------------

The following are base command arguments for the CLI:

    .. code-block:: bash

        imap-data-access -h # or
        imap-data-access query # or
        imap-data-access download # or
        imap-data-access upload

Add the -h flag with any base command for more information on use and functionality.

.. openapi:: openapi.yml
   :group:

>    :include: /query

Reviewer: This part is describing the REST API specification, so it should be moved to under that section.

**Example Usage:**

.. code-block:: bash

    imap-data-access query --start-date 20240101 --end-date 20241231 --output-format json
    # The following line is returned:
    [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}]

.. openapi:: openapi.yml
   :group:

>    :include: /download

Reviewer: Similar to the last comment, this is describing the REST API so should also be moved (same with the `/upload` section.

The API can be accessed from the following URL [WIP]: https://api.dev.imap-mission.com

Command Line Utility
--------------------
To Install
^^^^^^^^^^

Run the following command to use the API CLI:

    .. code-block:: bash

        pip install imap-data-access

Base Command Arguments
^^^^^^^^^^^^^^^^^^^^^^

The following are base command arguments for the CLI:

    .. code-block:: bash

        imap-data-access -h # or
        imap-data-access query # or
        imap-data-access download # or
        imap-data-access upload

Add the -h flag with any base command for more information on use and functionality.

Query
^^^^^

To query for files, you can use several parameters: ``--instrument``, ``--data-level``, ``--descriptor``, etc.

Further information is found in in the ``query -h`` menu. You can use parameters alone, or in combination.

**Example Usage:**

    .. code-block:: bash

        imap-data-access query --start-date 20240101 --end-date 20241231 --output-format json
        # The following line is returned:

>         [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}]

Reviewer: Could you split this up into multiple lines? The way this is displayed on the page (one long line that you have to scroll through) is a bit difficult to see.

Download
^^^^^^^^

> To download files using the CLI tool, use the command ``download``. These will be placed in a ``data`` directory.

Reviewer: ```suggestion
To download files using the CLI tool, use the command ``download``. The downloaded files will be placed in a ``data`` directory.
```

It is important to note that your working directory will be established as the default directory. I.e, the ``data``
directory will automatically be placed in this file path. Choose your working directory
accordingly to suit your desires.

When downloading a file from the API, different folders within the ``data`` directory will be made to better

> organize the downloaded files. See the example path: ``data/imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_20240105_v00-01.pkts``

Reviewer: ```suggestion
organize the downloaded files. See the example path: ``data/imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_20240105_v00-01.pkts``. The ``data`` directory and its structure is futher described [here](<link to data directory section>)
```

    .. code-block:: bash

        imap-data-access download imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts

Upload
^^^^^^

Similarly, files can be uploaded to the API using the command ``upload``.

When uploading files to the API, ensure these files are stored properly in a ``data`` directory (see the Data Directory section below for more information). Then,
ensure your working directory is one level above ``data`` in order to properly upload files.

[WIP] Certain ancillary files can also be uploaded to the API. For more specific information regarding these files, visit
`Ancillary Files <https://imap-processing.readthedocs.io/en/latest/data-access-api/calibration-files.html>`_

**Example Usage:**

    .. code-block:: bash

        imap-data-access upload /imap/swe/l1a/2024/01/imap_swe_l1a_sci_20240105_v001.cdf

Importing as a package
----------------------
Imap data access can also be imported and used as a python package if desired.

**Example Usage:**

    .. code-block:: bash

        import imap_data_access

        # Search for files
        results = imap_data_access.query(instrument="mag", data_level="l0")
        # results is a list of dictionaries

>         # [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105','version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}]

Reviewer: Similar to my comment above, if this could be split into multiple lines I think that would read better

        # Download a file that was returned from the search
        imap_data_access.download("imap/mag/l0/2024/01/imap_mag_l0_raw_202040101_v001.pkts")

        # Upload a calibration file that exists locally
        imap_data_access.upload("imap/swe/l1a/2024/01/imap_swe_l1a_sci_20240105_v001.cdf")

Configuration
--------------
Data Directory
^^^^^^^^^^^^^^

The folder structure for data files within the IMAP SDC is rigidly defined, so the data access will mimic that structure to make sure all data is stored in the same hierarchical structure as the SDC. This will enable seamless transition between a user's local system and the SDC. This is only used for downloads.

> A user's root data location can be specified as an environment variable IMAP_DATA_DIR or through a configuration dictionary within the package itself imap_data_access.config["DATA_DIR"]. If the IMAP_DATA_DIR variable is not set, the program defaults to the user's current working directory + data/.

Reviewer: ```suggestion
A user's root data location can be specified as an environment variable ``IMAP_DATA_DIR`` or through a configuration dictionary within the package itself (``imap_data_access.config["DATA_DIR"]``). If the ``IMAP_DATA_DIR`` variable is not set, the program defaults to the user's current working directory + ``data/``.
```

The following is the directory structure the IMAP SDC uses.

    .. code-block:: bash

        <IMAP_DATA_DIR>/
          imap/
            <instrument>/
              <data_level>/
                <year>/
                  <month>/
                    <filename>

for example, with ``IMAP_DATA_DIR=/data:``

    .. code-block:: bash

        /data/
          imap/
            swe/
              l0/
                2024/
                  01/
                    imap_swe_l0_sci_20240105_v001.pkts

Data Access URL
^^^^^^^^^^^^^^^

> To change the default URL that the package accesses, you can set the environment variable ``IMAP_DATA_ACCESS_URL`` or within the package ``imap_data_access.config["DATA_ACCESS_URL"]``. The default is the development server https://api.dev.imap-mission.com.

Reviewer: ```suggestion
To change the default URL that the package accesses, you can set the environment variable ``IMAP_DATA_ACCESS_URL`` or within the package ``imap_data_access.config["DATA_ACCESS_URL"]``. The default is the development server (``https://api.dev.imap-mission.com``).
```

File Validation
---------------

> This package validates filenames and paths to check they follow our standards, as defined by the filename conventions. There is also a class available for use by other packages to create filepaths and filenames that follow the IMAP SDC conventions.

Reviewer: ```suggestion
This package validates filenames and paths to check they follow our standards, as defined by the [filename conventions](https://imap-processing.readthedocs.io/en/latest/development-guide/style-guide/naming-conventions.html). There is also a class available for use by other packages to create filepaths and filenames that follow the IMAP SDC conventions.
```

The API can be accessed from the following URL [WIP]: https://api.dev.imap-mission.com

> To Install
> ==========

Reviewer: ```suggestion
To Install
----------
```

I think these section headers need to be changed to 'subsections'. I notice that the way these are currently defined is messing up the Table of Contents on the main landing page.
